kolibri/logger/tasks.py

```diff
@@ -0,0 +1,87 @@
+import logging as logger
+import threading
+import time
+
+from django.db import transaction
+
+logging = logger.getLogger(__name__)
+
+
+class AsyncLogQueue():
+
+    def __init__(self):
+
+        # Value in seconds to determine the sleep time between log saving batches
+        self.log_saving_interval = 5
+
+        # Where new log saving functions are appended
+        self.queue = []
+
+        # Where the to be executed log saving functions are stored
+        # once a batch save has been invoked
+        self.running = []
+
+    def append(self, fn):
+        """
+        Convenience method to append log saving function to the current queue
+        """
+        self.queue.append(fn)
+
+    def toggle_queue(self):
+        """
+        Method to swap the queue and running, to allow new log saving functions
+        to be added to the queue while previously added functions are being executed
+        and cleared without fear of race conditions dropping saves.
+        """
+        old_queue = self.queue
+        new_queue = self.running
+        self.queue = new_queue
+        self.running = old_queue
+
+    def clear_running(self):
+        """
+        Reset the running list to drop references to already executed log saving functions
+        """
+        self.running = []
+
+    def run(self):
+        """
+        Execute any log saving functions in the self.running list
+        """
+        if self.running:
+            # Do this conditionally to avoid opening an unnecessary transaction
+            with transaction.atomic():
+                for fn in self.running:
+                    try:
+                        fn()
```

Comment:
I think it's more efficient to save all the new logs through a bulk create method:

```
Log.objects.create(obj1, obj2, obj3)
```

Rather than having each model be saved individually: ` Log.objects.create(obj1), Log.objects.create(obj2), etc.`

Since it allows the Django ORM to use bulk create SQL queries.

Reply:
I am only using these for updates, not for creates, and morango has special sauce in its save method that would get skipped by a bulk update.

Also, the models being saved can be heterogeneous!

Reply:
Note that we currently *cannot* do bulk updates/saves because then the morango updates that are required (setting dirty bits etc) would not get done.

Also, wrapping multiple SQL statements in a single transaction is significantly more efficient than opening a new transaction for each statement (which is what would happen without this batch writing). c.f. https://medium.com/@JasonWyatt/squeezing-performance-from-sqlite-insertions-971aff98eef2

Reply:
I think this will benefit the performance because the enqueued task is releasing the server thread of these tasks. The response will be faster and more requests per second can be attended.
As Devon says, the database will have the same impact, but avoiding writing locks in the server thread loop will give to the user a better experience. The database writing queue will be the same, but it should not affect so much to the clients.

Reply:
Yeah, that's definitely also the case!

Reply:
The main concern might be that `POST` operations (as this only affects updates) might be slowed down by the background log updating, which could cause issues.